Sources/ArgumentParser/Completions/ZshCompletionsGenerator.swift

```diff
@@ -9,217 +9,236 @@
 //
 //===----------------------------------------------------------------------===//
 
-struct ZshCompletionsGenerator {
+extension [ParsableCommand.Type] {
   /// Generates a Zsh completion script for the given command.
-  static func generateCompletionScript(_ type: ParsableCommand.Type) -> String {
-    let initialFunctionName = [type].completionFunctionName()
+  var zshCompletionScript: String {
+    """
+    #compdef \(first?._commandName ?? "")
+
+    \(completeFunctionName)() {
+        local -ar non_empty_completions=("${@:#(|:*)}")
+        local -ar empty_completions=("${(M)@:#(|:*)}")
+        _describe '' non_empty_completions -- empty_completions -P $'\\'\\''
+    }
 
-    return """
-      #compdef \(type._commandName)
-      local context state state_descr line
-      _\(type._commandName.zshEscapingCommandName())_commandname=$words[1]
-      typeset -A opt_args
-
-      \(generateCompletionFunction([type]))
-      _custom_completion() {
-          local completions=("${(@f)$($*)}")
-          _describe '' completions
-      }
+    \(customCompleteFunctionName)() {
+        local -a completions
+        completions=("${(@f)"$("${@}")"}")
+        if [[ "${#completions[@]}" -gt 1 ]]; then
+            \(completeFunctionName) "${completions[@]:0:-1}"
+        fi
+    }
 
-      \(initialFunctionName)
-      """
+    \(completionFunctions)\
+    \(completionFunctionName())
+    """
   }
 
-  static func generateCompletionFunction(_ commands: [ParsableCommand.Type])
-    -> String
-  {
-    guard let type = commands.last else { return "" }
-    let functionName = commands.completionFunctionName()
-    let isRootCommand = commands.count == 1
+  private var completionFunctions: String {
+    guard let type = last else { return "" }
+    let functionName = completionFunctionName()
+    let isRootCommand = count == 1
+
+    let argumentSpecsAndSetupScripts = argumentsForHelp(visibility: .default)
```

Comment:
I know this matches what we're doing now, but perhaps we should include `.hidden` arguments in the completion script, just not `.private`. Let's discuss later…

Reply:
I don't know the exact difference between `.hidden` & `.private`, but, making an educated guess, that sounds like it could make sense.

Since I don't know SAP very well, I tried to limit my changes to fixing shell issues & fixing SAP interactions that I know are broken, but it's a great idea to revisit existing assumptions throughout the completion script generation process.

I assume we should leave this particular improvement for a subsequent PR so it can be mulled over & discussed, but I'm happy to include it here if you so instruct me.

Reply:
This is definitely for later, just something I noticed when reviewing… `.hidden` args aren't included in the default `--help` output, but are visible with `--help-hidden`. `.private` args are never included.